example/contrib/socks-client/socks/impl/handshake.hpp

```diff
@@ -0,0 +1,598 @@
+//
+// Copyright (c) 2018 jackarain (jack dot wgm at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/beast
+//
+
+#ifndef SOCKS_IMPL_HANDSHAKE_HPP
+#define SOCKS_IMPL_HANDSHAKE_HPP
+
+#include <socks/detail/protocol.hpp>
+#include <boost/beast/core/async_base.hpp>
+#include <boost/beast/core/detail/is_invocable.hpp>
+#include <boost/config.hpp> // for BOOST_FALLTHROUGH
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/asio/streambuf.hpp>
+#include <boost/asio/read.hpp>
+#include <boost/asio/write.hpp>
+#include <boost/asio/buffer.hpp>
+#include <iostream>
+#include <string>
+#include <memory>
+#include <utility>
+#include <type_traits>
+
+
+namespace socks {
+
+namespace detail {
+
+template<typename type, typename source>
+type read(source& p)
+{
+    type ret = 0;
+    for (std::size_t i = 0; i < sizeof(type); i++)
+        ret = (ret << 8) | (static_cast<unsigned char>(*p++));
+    return ret;
+}
+
+template<typename type, typename target>
+void write(type v, target& p)
+{
+    for (auto i = (int)sizeof(type) - 1; i >= 0; i--, p++)
+        *p = static_cast<unsigned char>((v >> (i * 8)) & 0xff);
+}
+
+} // detail
+
+template<
+    class Stream,
+    class Handler,
+    class Buffer,
+    class base_type = boost::beast::async_base<
+        Handler, typename Stream::executor_type>>
+class socks4_op : public base_type
```

Comment:
`socks4_op` belongs in the `detail` namespace since it is not a public interface